File: Submariner/SBRepeatModeTransformer.m
//
//  SBRepeatModeTransformer.m
//  Submariner
//
//  Created by Calvin Buckley on 2022-05-23.
//  Copyright © 2022 Calvin Buckley. All rights reserved.
//

#import "SBRepeatModeTransformer.h"

@implementation SBRepeatModeTransformer {
    SBPlayerRepeatMode mode;
}

- (id) initWithMode: (SBPlayerRepeatMode) newMode {
    if (!(self = [super init])) return nil;
    mode = newMode;
    return self;
}

+ (Class)transformedValueClass {
    return [NSNumber class];
}

// Makes menu item happy
+ (BOOL)allowsReverseTransformation {
    return NO;
}

- (id)transformedValue:(id)value {
    SBPlayerRepeatMode passedMode = (SBPlayerRepeatMode)[value integerValue];
    return @(mode == passedMode);
}

- (id)reverseTransformedValue:(id)value {
    BOOL enabled = [value boolValue];
    NSLog(@"Enabled? %d; Transformer Mode? %d", enabled, mode);
    return enabled ? @(mode) : @(SBPlayerRepeatNo);
}

@end

File: Submariner/SBRepeatModeTransformer.swift
//
//  SBRepeatModeTransformer.swift
//  Submariner
//
//  Created by Calvin Buckley on 2023-04-10.
//  Copyright © 2023 Submariner Developers. All rights reserved.
//

import Cocoa

@objcMembers class SBRepeatModeTransformer: ValueTransformer {
    let newMode: SBPlayerRepeatMode
    
    init(mode: SBPlayerRepeatMode) {
        self.newMode = mode
        super.init()
    }
    
    override class func allowsReverseTransformation() -> Bool {
        true
    }
    
    override class func transformedValueClass() -> AnyClass {
        return NSNumber.self
    }
    
    override func transformedValue(_ value: Any?) -> Any? {
        if let modeRaw = value as? Int, let passedMode = SBPlayerRepeatMode(rawValue: modeRaw) {
            return self.newMode == passedMode
        }
        return nil
    }
    
    override func reverseTransformedValue(_ value: Any?) -> Any? {
        if let enabled = value as? Bool {
            let ret = enabled ? newMode : .no
            return NSNumber.init(integerLiteral: ret.rawValue)
        }
        return nil
    }
}
